Make SurfaceTool.generate_normals() behave consistently with smoothing groups

Klowner opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->
I'm not entirely sure how smooth groups is *supposed* to work within SurfaceTool since it seems to be undocumented, but I think this relatively small change makes the behavior easier to reason about and produce results in a consistent manner. 

This change addresses a couple problems(?) with `generate_normals()`. The first issue is that vertices are binned using a hash map which takes into account the *entire* vertex (including uvs, colors, bones, etc.). To address this issue, this PR introduces a `SmoothGroupVertex` struct and accompanying `SmoothGroupVertexHasher` which only takes vertex position and smooth group into account. This means that vertices of adjacent faces sharing the same smooth group will *contribute to* and *receive* smoothed normals regardless of other vertex attributes.

The other potential issue that some other tickets (see below) bring up is that for a user to generate normals for a mesh with flat faces it would be necessary to use different smooth groups for each face (or at least ensuring that no adjacent vertices share smooth groups) which seems to me like an unnecessary burden for the user. This PR also includes changes which imply that surface group `0` is meant to be flat shaded. If a user wishes to create a smooth shaded mesh, a single call to `surface_tool.set_smooth_group(1)` would mean the entire mesh will be smooth. This behavior makes sense to me, but I defer to the group :grin: .

Here's a screenshot of the output of the attached example project (adapted from a related issue below):
![generate_normals_changes](https://user-images.githubusercontent.com/9374/198852167-e4375ad2-f1fd-495b-96ee-d7c407750ec7.png)
Example:
[SurfaceToolSmoothShading4.zip](https://github.com/godotengine/godot/files/9894596/SurfaceToolSmoothShading4.zip)

Related issues:
 - #6146 - (2.x and 3.x) describes an issue where smoothing is not performed when vertices of adjacent faces don't share UV coordinates (binning issue)
 - #65076 - (4.0) - claims 3.x works fine, but the example consists of flat sided cubes which masks the issue, while the problem is clearly visible in the 4.x screenshot (also a binning issue)
 - #65104 - (3.5) - identifies part of the vertex binning issue, but 3.x's `add_smooth_group()` has a slightly unrelated bug in addition to the binning.

Although this PR is not applicable to 3.x, I think similar behavior can be implemented with the old `add_smooth_group(bool)` method.

Hopefully this is a welcome contribution, any feedback is greatly appreciated!

*Bugsquad edit:*
- Fixes https://github.com/godotengine/godot/issues/6146
- Fixes https://github.com/godotengine/godot/issues/65076
- Fixes https://github.com/godotengine/godot/issues/65104
- Fixes https://github.com/godotengine/godot/issues/66765

Klowner committed:
Make SurfaceTool.generate_normals() behave in a consistent manner in regard to smoothing groups, imply group 0 is flat

kleonc commented:
Can't tell if that's the best / preferred approach but the justification makes sense. The changes should probably be fine as long as the new behavior is documented. So yeah, please update the docs (`SurfaceTool.xml`) according to the changes made in this PR as well.

cc @clayjohn? (kinda related to rendering)

Klowner commented:
Added docs!

pseidemann commented:
hi,
nice change! I just ran into this as well.
does this change also fix a potential bug that UVs are considered when comparing vertices when searching for duplicates e.g. in index()?
right now I think there is a related bug that once I set different UVs for vertices, not only set_smooth_group() does not work properly but also index() and I actually think also commit() will generate more vertices than needed because of this.

might be worth fixing the overall definition of when two vertices are considered identical/redundant?

pseidemann commented:
is there any way to smoothen the normals after committing in go 3.5 maybe with MeshDataTool?

pseidemann commented:
after some fiddling I was able to write a function which computes smooth normals for a given mesh (should work with any godot version):

```gdscript
func smooth_normals(input_mesh, output_mesh):
	var mdt = MeshDataTool.new()

	# read input mesh
	mdt.create_from_surface(input_mesh, 0)

	# group vertices by position
	var grouped = {}
	for i in range(mdt.get_vertex_count()):
		var pos = mdt.get_vertex(i)
		if !grouped.has(pos):
			grouped[pos] = []
		grouped[pos].push_back(i)

	# process every distinct position only once
	for pos in grouped:
		var vertex_ids = grouped[pos]
		var adjacent_faces = []

		# iterate all vertices with the same position
		for i in vertex_ids:
			var faces = mdt.get_vertex_faces(i)
			# put all faces of all vertices with the same position in one list
			adjacent_faces.append_array(faces)

		# get sum of all face normals of all adjacent faces of current position
		var normal = Vector3.ZERO
		for i in adjacent_faces:
			normal += mdt.get_face_normal(i)
		normal = normal.normalized()

		# set computed normal for all vertices with same position
		for i in vertex_ids:
			mdt.set_vertex_normal(i, normal)

	# write to output mesh
	mdt.commit_to_surface(output_mesh)
```

Klowner commented:
@pseidemann yep, this change fixes the issue where smoothing isn't performed if UVs are incongruous.

Considering how indexed rendering works, `index()` creating additional vertices when setting UVs isn't so much a bug as it would be expected behavior, no?

I'm also happy to remove the `0` indicating flat shading from this PR.

pseidemann commented:
@Klowner 
> Considering how indexed rendering works, `index()` creating additional vertices when setting UVs isn't so much a bug as it would be expected behavior, no?

true

clayjohn commented:
> I'm also happy to remove the `0` indicating flat shading from this PR.

Without that change this PR is a bugfix and we could go ahead and merge it without @reduz's review. 

That being said, I agree that needing to specify a unique smooth group per vertex to get flat shaded normals is awkward at best. So it may be worthwhile to wait for his input so we can improve the behaviour while we are at it

clayjohn commented:
We discussed this on the rendering Rocketchat and agreed that -1 would be a better value for specifying flat normals. I rebased this PR and made the change in this commit https://github.com/clayjohn/godot/commit/4f26b573ce0b1f8ebd5b80b0a6b95c62274a6d76

@Klowner Will you have a chance to incorporate this into your PR sometime this week? If not, I can make a PR with your changes and list you as the co-author.

Klowner commented:
Added your change, @clayjohn. Also rebased atop master and incorporated the `for (v: vectors)` iterator syntax (which I assume is the new preferred style). :man_dancing:

clayjohn reviewed (approved):
Thank you for updating this so quickly! Looks good to me

clayjohn commented:
The final step before merging is to squash all the commits together so that the whole PR only contains 1 big commit with all your changes. We like to merge one commit at a time to keep the git history clean and navigable. 

If you don't know how to do that, we have a helpful tutorial in the official documentation https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase

Klowner commented:
Ah, thank you, flat as a pancake :+1:

akien-mga commented:
Thanks!

akien-mga commented:
> Related issues:
>
> - https://github.com/godotengine/godot/issues/6146 - (2.x and 3.x) describes an issue where smoothing is not performed when vertices of adjacent faces don't share UV coordinates (binning issue)
> - https://github.com/godotengine/godot/issues/65076 - (4.0) - claims 3.x works fine, but the example consists of flat sided cubes which masks the issue, while the problem is clearly visible in the 4.x screenshot (also a binning issue)
> - https://github.com/godotengine/godot/issues/65104 - (3.5) - identifies part of the vertex binning issue, but 3.x's add_smooth_group() has a slightly unrelated bug in addition to the binning.

Are those 3 fixed by this PR?

Klowner commented:
Yep! fixes #6146, fixes #65076, and fixes #65104.

Klowner commented:
Fixes #66765 as well, since it should now be possible to disable smoothing.
